File: backend/app/schemas/learning.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class LearningVideoRead(ORMModel):
    id: int
    title: str
    description: str | None = None
    original_filename: str
    storage_key: str
    content_type: str
    file_size_bytes: int
    duration_seconds: int
    completion_threshold_percent: int
    status: str
    uploaded_at: datetime
    created_at: datetime
    updated_at: datetime
    playback_url: str


class LearningVideoUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None


class LearningVideoProgressRead(BaseModel):
    last_position_seconds: int = 0
    watched_seconds: int = 0
    completion_percent: int = 0
    completed_at: datetime | None = None
    last_heartbeat_at: datetime | None = None


class CandidateLearningVideoRead(LearningVideoRead):
    progress: LearningVideoProgressRead


class LearningProgressUpdate(BaseModel):
    current_position_seconds: int = Field(ge=0)
    watched_start_seconds: int = Field(ge=0)
    watched_end_seconds: int = Field(ge=0)


class LearningReportRow(BaseModel):
    candidate_id: int
    candidate_name: str
    employee_no: str | None = None
    department: str | None = None
    exam_group: str | None = None
    video_id: int
    video_title: str
    video_status: str
    duration_seconds: int
    completion_percent: int
    completion_status: str
    last_heartbeat_at: datetime | None = None
    completed_at: datetime | None = None

File: frontend/src/types/learning.ts
export type LearningVideoStatus = "draft" | "published" | "archived";
export type LearningCompletionStatus = "not_started" | "in_progress" | "completed";

export type LearningVideoProgress = {
  last_position_seconds: number;
  watched_seconds: number;
  completion_percent: number;
  completed_at?: string | null;
  last_heartbeat_at?: string | null;
};

export type LearningVideo = {
  id: number;
  title: string;
  description?: string | null;
  original_filename: string;
  storage_key: string;
  content_type: string;
  file_size_bytes: number;
  duration_seconds: number;
  completion_threshold_percent: number;
  status: LearningVideoStatus;
  uploaded_at: string;
  created_at: string;
  updated_at: string;
  playback_url: string;
};

export type CandidateLearningVideo = LearningVideo & {
  progress: LearningVideoProgress;
};

export type LearningProgressPayload = {
  current_position_seconds: number;
  watched_start_seconds: number;
  watched_end_seconds: number;
};

export type LearningVideoUpdatePayload = {
  title?: string;
  description?: string | null;
};

export type LearningVideoUploadPayload = {
  title: string;
  description?: string | null;
  duration_seconds: number;
  file: File;
};

export type LearningReportRow = {
  candidate_id: number;
  account_email: string;
  display_name: string;
  account_status: "pending" | "active" | "inactive" | string;
  video_id: number;
  video_title: string;
  video_status: LearningVideoStatus;
  duration_seconds: number;
  completion_percent: number;
  completion_status: LearningCompletionStatus;
  last_heartbeat_at?: string | null;
  completed_at?: string | null;
};
